File: src/main/kotlin/school/charset/app/domain/exercise/generator/Utf32ExerciseGenerator.kt
package school.charset.app.domain.exercise.generator

import school.charset.app.domain.encoding.Codec
import school.charset.app.domain.encoding.Encoding
import school.charset.app.domain.exercise.Exercise
import school.charset.app.domain.exercise.ExerciseGenerationException

class Utf32ExerciseGenerator(
    private val codec: Codec,
    override val encoding: Encoding,
    private val codePointGenerator: CodePointGenerator,
    private val byteArrayGenerator: ByteArrayGenerator,
    private val utf32Generator: Utf32Generator,
) : EncodingExerciseGenerator {

    init {
        require(encoding == Encoding.Utf32Be || encoding == Encoding.Utf32Le) {
            "Utf32ExerciseGenerator requires Utf32Be or Utf32Le, got ${encoding.id}"
        }
    }

    private val endian: Encoding.Endian =
        if (encoding == Encoding.Utf32Be) Encoding.Endian.BigEndian else Encoding.Endian.LittleEndian

    override fun generateEncode(level: Int): Exercise.Encode {
        val utf32Level = parseLevel(level)
        val codePoint = codePointGenerator.randomUtf32(utf32Level)
        val steps = utf32Generator.buildEncodeStepsFor(codePoint, endian)
        return Exercise.Encode(codePoint, encoding, level, steps)
    }

    override fun generateDecode(level: Int): Exercise.Decode {
        val utf32Level = parseLevel(level)
        val bytes = byteArrayGenerator.randomUtf32(utf32Level, encoding)
        val codePoint = codec.decode(bytes, encoding)
        val steps = utf32Generator.buildDecodeStepsFor(bytes, codePoint, endian)
        return Exercise.Decode(bytes, codePoint, encoding, level, steps)
    }

    private fun parseLevel(level: Int): Utf32Level = Utf32Level.fromNumber(level)
        ?: throw ExerciseGenerationException(
            encoding = encoding,
            level = level,
            reason = "level must be one of: ${Utf32Level.validNumbers}",
        )
}

File: lib/charset/exercise/generator/utf32_exercise_generator.ex
defmodule Charset.Exercise.Generator.Utf32ExerciseGenerator do
  @moduledoc """
  UTF-32 exercises (BE or LE). The endianness step is dropped - the exercise
  gives the target byte order in its header (a random BE/LE), it isn't
  something the learner derives. Byte order is still tested via the
  hex-bytes step. The sandbox keeps the endianness step (shared `Utf32Steps`)
  for its explanation panel.
  """

  alias Charset.Encoding
  alias Charset.Encoding.Codec
  alias Charset.Exercise
  alias Charset.Exercise.Generator.CodePointGenerator
  alias Charset.Exercise.Generator.Levels
  alias Charset.Exercise.Generator.Utf32Steps
  alias Charset.Exercise.GenerationError
  alias Charset.Exercise.Step

  @spec generate_encode!(:utf32be | :utf32le, pos_integer()) :: Exercise.t()
  def generate_encode!(encoding, level) when encoding in [:utf32be, :utf32le] do
    tier = parse_level!(encoding, level)
    code_point = CodePointGenerator.random_utf32(tier)

    Exercise.encode(code_point, encoding, level, exercise_encode_steps(code_point, encoding))
  end

  @spec generate_decode!(:utf32be | :utf32le, pos_integer()) :: Exercise.t()
  def generate_decode!(encoding, level) when encoding in [:utf32be, :utf32le] do
    tier = parse_level!(encoding, level)
    bytes = Codec.encode!(CodePointGenerator.random_utf32(tier), encoding)
    code_point = Codec.decode!(bytes, encoding)

    Exercise.decode(
      bytes,
      code_point,
      encoding,
      level,
      exercise_decode_steps(bytes, code_point, encoding)
    )
  end

  @doc "Deterministic exercise-flow step builder (exposed for tests)."
  @spec exercise_encode_steps(integer(), :utf32be | :utf32le) :: [Step.t()]
  def exercise_encode_steps(code_point, encoding) do
    code_point
    |> Utf32Steps.build_encode_steps_for(Encoding.endianness(encoding))
    |> without_endianness()
  end

  @doc "Deterministic exercise-flow step builder (exposed for tests)."
  @spec exercise_decode_steps(binary(), integer(), :utf32be | :utf32le) :: [Step.t()]
  def exercise_decode_steps(bytes, code_point, encoding) do
    bytes
    |> Utf32Steps.build_decode_steps_for(code_point, Encoding.endianness(encoding))
    |> without_endianness()
  end

  defp without_endianness(steps) do
    Enum.reject(steps, &match?(%Step.Endianness{}, &1))
  end

  defp parse_level!(encoding, level) do
    case Levels.parse(:utf32, level) do
      {:ok, tier} ->
        tier

      :error ->
        raise GenerationError,
          encoding: encoding,
          level: level,
          reason: "level must be one of: #{Levels.valid_numbers(:utf32)}"
    end
  end
end
